File: lower_common_ancestor_bst.py
# Time Complexity: O(h)
# Space Complexity: O(h)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
Given that we are working with a BST and we are given two nodes p and q, we can use the property of a BST 
where smaller or larger number will be on one side of the node. We continue doing this until this is not 
true and we return the node we are on.
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root.val < p.val and root.val < q.val:
            return self.lowestCommonAncestor(root.right, p, q)
        elif root.val > p.val and root.val > q.val:
            return self.lowestCommonAncestor(root.left, p, q)
        else:
            return root

File: LowerCommonAncestorBST.java
//TC will be O(h), where h is the height of the tree
//SC will be O(h)

//USing recursive approach

//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class SolutLowerCommonAncestorBST {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null){
            return root;
        }

        return lca(root, p, q);
    }

    private TreeNode lca(TreeNode root, TreeNode p, TreeNode q){
        //base
        if(root == null){
            return null;
        }
        //logic
        if(root.val > p.val && root.val > q.val){
            return lca(root.left, p, q);
        }
        else if(root.val < p.val && root.val< q.val){
            return lca(root.right, p, q);
        }
        else{
            return root;
        }
    }

    public static void main(String[] args) {
        // Create the BST using the correct data structure
        TreeNode root = new TreeNode(6);
        root.left = new TreeNode(2);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(0);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(7);
        root.right.right = new TreeNode(9);
        root.left.right.left = new TreeNode(3);
        root.left.right.right = new TreeNode(5);

        TreeNode p = root.left; // Node with value 2
        TreeNode q = root.right; // Node with value 8

        SolutLowerCommonAncestorBST obj = new SolutLowerCommonAncestorBST();
        TreeNode result = obj.lowestCommonAncestor(root, p, q);

        System.out.println("Lowest Common Ancestor: " + result.val);
    }
}